### ISSUE TITLE
Handling of no/zero gradients in cpp Function

### ISSUE THREAD

**Issue by albanD** [pytorch/pytorch#2003]
The problem here is what should be the returned value of a cpp `Function` when the output was indépendant of the inputs or no gradient was backpropagated for this specific output (it was unused later in the graph).

For python `Function`, the wrapper takes care, when no gradient exist for an output to create a full dense tensor filled with `0`s of the size of the output. The same happens when a python `Function` returns `None` for a given parameter, the next `Function` will actually see a dense tensor of `0`s.

In cpp, such wrapper does not exist, so if an output was not used, the corresponding element in the input `variable_list` will be and empty `shared_ptr`. And the cpp `Function` should handle empty `shared_ptr` as if it was a dense Tensor of `0`s (possibly skipping a lot of computations as in `ConvBackwardBackward`).
Right now, some `Function` do handle this case properly, but not all of them (leading for example to [this](https://discuss.pytorch.org/t/error-using-autograd-grad-with-2-directly-following-conv2d-layers/4661) issue in the forum).

The question is: are we ok with the above design saying that any input which is an empty `shared_ptr` is equivalent to a tensor of all `0`s ? And all cpp `Function`s should handle empty `shared_ptr` as input ?

**Comment by apaszke** [comment]
Yes. Empty `shared_ptr`s should be treated as all-zero tensors. We pre-allocate them for Python functions, so it's easier to implement them for the users, but we should treat them specially in case of C++ functions. We could add a tests that starts backward with `None` to make sure they don't segfault.

**Comment by albanD** [comment]
Ok,

One question I would have is for simple functions that are not implemented in pure autograd like [this](https://github.com/pytorch/pytorch/blob/ebdec9a837074a303fd5ffb6f319cd593955becc/torch/csrc/autograd/functions/tensor.cpp#L14-L24) one. what should we do?
Should we adapt the `wrap_outputs` function such that if a tensor does not exist, it places an empty `shared_ptr` in the `variable_list` ?
If this goes back to python side, this will be properly translated to a `None` so that the python will work as expected.

**Comment by apaszke** [comment]
What do you mean by functions that are not implemented in pure autograd? All of C++ functions?
Yeah, empty tensor `unique_ptr`s should be translated to empty variable `shared_ptr`s

**Comment by albanD** [comment]
Sorry, for me cpp `Function`s not implemented in pure autograd means the ones that do not use `wrap_outputs`.

**Comment by apaszke** [comment]
I'm closing this, because it's more of a question than an issue. We can always discuss this on Slack!
